File: Labs/Lab 6/C#/Row6/Text.cs
using System.Text;

namespace TextLib;

public class Text : IWhitespaceCleaner
{
    private readonly List<Row> _rows = new();

    // --- IWhitespaceCleaner ---
    public string NormalizeSpaces(string s)
    {
        var sb = new StringBuilder();
        bool prevSpace = false;
        foreach (char ch in s)
        {
            if (char.IsWhiteSpace(ch))
            {
                if (!prevSpace) sb.Append(' ');
                prevSpace = true;
            }
            else
            {
                sb.Append(ch);
                prevSpace = false;
            }
        }
        return sb.ToString();
    }

    public string Trim(string s) => s.Trim();

    // --- керування рядками ---
    public void AddRow(Row r) => _rows.Add(r);
    public bool RemoveRow(int index)
    {
        if (index < 0 || index >= _rows.Count) return false;
        _rows.RemoveAt(index);
        return true;
    }
    public void Clear() => _rows.Clear();

    // --- статистика ---
    public int ShortestRowLength() =>
        _rows.Count == 0 ? 0 : _rows.Min(r => r.Length);

    public double ConsonantPercentage()
    {
        int total = 0, consonants = 0;
        foreach (var r in _rows)
        {
            string s = r.Data;
            total += s.Count(char.IsLetter);
            consonants += Row.CountConsonants(s);
        }
        return total == 0 ? 0 : 100.0 * consonants / total;
    }
}

File: Labs/Lab 6/C++/Rows6/Text.cpp
#include "Text.h"
#include <sstream>
#include <algorithm>
#include <cctype>

std::string Text::NormalizeSpaces(const std::string& s) const {
    std::ostringstream out;
    bool prevSpace = false;
    for (char ch : s) {
        if (std::isspace(static_cast<unsigned char>(ch))) {
            if (!prevSpace) { out << ' '; }
            prevSpace = true;
        }
        else {
            out << ch;
            prevSpace = false;
        }
    }
    return out.str();
}

std::string Text::Trim(const std::string& s) const {
    size_t start = 0, end = s.size();
    while (start < end && std::isspace(static_cast<unsigned char>(s[start]))) ++start;
    while (end > start && std::isspace(static_cast<unsigned char>(s[end - 1]))) --end;
    return s.substr(start, end - start);
}

void Text::AddRow(const Row& r) { rows.push_back(r); }

bool Text::RemoveRow(size_t index) {
    if (index >= rows.size()) return false;
    rows.erase(rows.begin() + index);
    return true;
}

void Text::Clear() { rows.clear(); }

size_t Text::ShortestRowLength() const {
    if (rows.empty()) return 0;
    size_t mn = rows.front().Length();
    for (const auto& r : rows) mn = std::min(mn, r.Length());
    return mn;
}

double Text::ConsonantPercentage() const {
    size_t totalLetters = 0, consonants = 0;
    for (const auto& r : rows) {
        const std::string& s = r.Get();
        totalLetters += std::count_if(s.begin(), s.end(), ::isalpha);
        consonants += Row::CountConsonants(s);
    }
    return totalLetters ? (100.0 * consonants / totalLetters) : 0.0;
}